File: OpenBid/DFP/ObjectiveC/SampleApp/SampleApp/EventHandlers/DFP/Banner/DFPBannerEventHandler.m
/*
 * PubMatic Inc. ("PubMatic") CONFIDENTIAL
 * Unpublished Copyright (c) 2006-2019 PubMatic, All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains the property of
 * PubMatic. The intellectual and technical concepts contained herein are
 * proprietary to PubMatic and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material is
 * strictly forbidden unless prior written permission is obtained from PubMatic.
 * Access to the source code contained herein is hereby forbidden to anyone
 * except current PubMatic employees, managers or contractors who have executed
 * Confidentiality and Non-disclosure agreements explicitly covering such
 * access.
 *
 * The copyright notice above does not evidence any actual or intended
 * publication or disclosure  of  this source code, which includes information
 * that is confidential and/or proprietary, and is a trade secret, of  PubMatic.
 * ANY REPRODUCTION, MODIFICATION, DISTRIBUTION, PUBLIC  PERFORMANCE, OR PUBLIC
 * DISPLAY OF OR THROUGH USE  OF THIS  SOURCE CODE  WITHOUT  THE EXPRESS WRITTEN
 * CONSENT OF PubMatic IS STRICTLY PROHIBITED, AND IN VIOLATION OF APPLICABLE
 * LAWS AND INTERNATIONAL TREATIES.  THE RECEIPT OR POSSESSION OF  THIS SOURCE
 * CODE AND/OR RELATED INFORMATION DOES NOT CONVEY OR IMPLY ANY RIGHTS TO
 * REPRODUCE, DISCLOSE OR DISTRIBUTE ITS CONTENTS, OR TO MANUFACTURE, USE, OR
 * SELL ANYTHING THAT IT  MAY DESCRIBE, IN WHOLE OR IN PART.
 */

#import "DFPBannerEventHandler.h"
#import <POBBid.h>

#define SYNC_TIMEOUT_INTEREVAL 0.6

@interface DFPBannerEventHandler () <GADBannerViewDelegate,
GADAppEventDelegate,
GADAdSizeDelegate>
@property(nonatomic, strong) DFPBannerView *bannerView;
@property(nonatomic) NSTimer *timer;
@property(nonatomic, weak) id<POBBannerEventDelegate> delegate;
@property(nonatomic) CGSize dfpAdSize;
@property(nonatomic, assign) BOOL notified;
@property(nonatomic, assign) BOOL isAppEventExpected;
@property(nonatomic, strong) NSArray *adSizes;
@end

@implementation DFPBannerEventHandler
- (instancetype)initAdunitId:(NSString *)adUnitId
                    andSizes:(NSArray *)validSizes {
    
    self = [super init];
    if (self) {
        
        // Create DFPBannerView and set ad unit and ad sizes
        _bannerView = [[DFPBannerView alloc] init];
        _bannerView.adUnitID = adUnitId;
        
        _adSizes = validSizes;
        [_bannerView setValidAdSizes:validSizes];
        _dfpAdSize = _bannerView.adSize.size;
        
        // Set delegates on DFPBannerView instance, these should not be removed/overridden else event handler will not work as expected.
        _bannerView.delegate = self;
        _bannerView.appEventDelegate = self;
        _bannerView.adSizeDelegate = self;
    }
    return self;
}

- (void)dealloc {
    _bannerView.delegate = nil;
    _bannerView = nil;
    _configBlock = nil;
}

#pragma mark - POBBannerEvent methods

- (void)setDelegate:(id<POBBannerEventDelegate>)delegate {
    _delegate = delegate;
}

// OpenBid SDK passes its bids through this method
- (void)requestAdWithBid:(POBBid *)bid {
    _notified = NO;
    _isAppEventExpected = NO;
    _bannerView.rootViewController = [self.delegate viewControllerForPresentingModal];
    
    // Create DFP ad request
    DFPRequest *dfpRequest = [[DFPRequest alloc] init];
    
    // Call configuration block if set. it can be used to configure DFP banner and ad request
    if (self.configBlock) {
        self.configBlock(_bannerView, dfpRequest);
    }
    
    if (!(_bannerView.appEventDelegate == self &&
        _bannerView.delegate == self)) {
        NSLog(@"Do not set DFP delegates. These are used by DFPBannerEventHandler internally.");
    }
    
    // If bid is valid, add bid related custom targetting on DFP ad request
    if (bid) {
        if (bid.status.boolValue) {
            _isAppEventExpected = YES;
        }
        
        NSMutableDictionary * customTargeting = [NSMutableDictionary dictionaryWithDictionary:dfpRequest.customTargeting];
        [customTargeting addEntriesFromDictionary:[bid targetingInfo]];
        [dfpRequest setCustomTargeting:[NSDictionary dictionaryWithDictionary:customTargeting]];
        NSLog(@"Custom targeting : %@", [customTargeting description]);
    }
    
    // Load ad request
    [_bannerView loadRequest:dfpRequest];
}

- (CGSize)adContentSize {
    return _dfpAdSize;
}

- (NSArray<POBAdSize *> *)requestedAdSizes {
    NSMutableArray *sizes = [NSMutableArray new];
    for (NSValue *adSize in _adSizes) {
        GADAdSize gAdSize = GADAdSizeFromNSValue(adSize);
        [sizes addObject:POBAdSizeMakeFromCGSize(gAdSize.size)];
    }
    return [NSArray arrayWithArray:sizes];
}

#pragma mark - GADAppEventDelegate methods

// Called when the banner receives an app event.
- (void)adView:(DFPBannerView *)banner
didReceiveAppEvent:(NSString *)name
      withInfo:(NSString *)info {
    if ([name isEqualToString:@"pubmaticdm"]) {
        if (_notified) {
            NSDictionary *userInfo = nil;
            NSString *errorMessage = @"App event is called unexpetedly";
            userInfo = @{NSLocalizedDescriptionKey : NSLocalizedString(errorMessage, nil)};
            NSError *error = [NSError errorWithDomain:kPOBErrorDomain
                                                 code:POBSignalingError
                                             userInfo:userInfo];
            [self.delegate failedWithError:error];
            return;
        }
        _notified = YES;
        [self.delegate openBidPartnerDidWin];
    }
}

#pragma mark - GADBannerViewDelegate methods

- (void)adViewDidReceiveAd:(GADBannerView *)bannerView {
    
    // If already notifed, skip waiting for app event
    if (_notified) {
        return;
    }
    
    // If OpenBid SDK have provided non-zero bid price, expect for app event for fixed time interval, otherwise consider as DFP has won & serving its own ad
    if (!_isAppEventExpected) {
        [self.delegate adServerDidWin:bannerView];
        self.notified = YES;
    } else {
        // Timer to synchronize did recieve and app event callback as their sequence is not fixed
        [self.timer invalidate];
        self.timer = [NSTimer scheduledTimerWithTimeInterval:SYNC_TIMEOUT_INTEREVAL target:self selector:@selector(syncTimetAction) userInfo:nil repeats:NO];
    }
}

/// Tells the delegate that an ad request failed. The failure is normally due to
/// network connectivity or ad availablility (i.e., no fill).
- (void)adView:(GADBannerView *)bannerView
didFailToReceiveAdWithError:(GADRequestError *)error {
    
    NSError *eventError = nil;
    switch (error.code) {
        case kGADErrorNoFill:
            eventError = [self eventError:error withErrorCode:POBErrorNoAds];
            break;
            
        case kGADErrorInvalidRequest:
            eventError = [self eventError:error withErrorCode:POBErrorInvalidRequest];
            break;
            
        case kGADErrorNetworkError:
            eventError = [self eventError:error withErrorCode:POBErrorNetworkError];
            break;
            
        case kGADErrorTimeout:
            eventError = [self eventError:error withErrorCode:POBErrorTimeout];
            break;
            
        case kGADErrorInternalError:
        case kGADErrorInterstitialAlreadyUsed:
        case kGADErrorMediationDataError:
        case kGADErrorMediationAdapterError:
        case kGADErrorMediationNoFill:
        case kGADErrorMediationInvalidAdSize:
        case kGADErrorInvalidArgument:
            eventError = [self eventError:error withErrorCode:POBErrorInternalError];
            break;
            
        case kGADErrorReceivedInvalidResponse:
            eventError = [self eventError:error withErrorCode:POBErrorInvalidResponse];
            break;
            
        default:
            eventError = error;
            break;
    }
    [_delegate failedWithError:eventError];
}

-(NSError *)eventError:(GADRequestError *)error  withErrorCode:(POBErrorCode )code{
    NSError *eventError = [NSError errorWithDomain:kPOBErrorDomain
                                         code:code
                                     userInfo:error.userInfo];
    return eventError;
}

- (void)adViewWillPresentScreen:(GADBannerView *)bannerView {
    [_delegate willPresentModal];
}

- (void)adViewDidDismissScreen:(GADBannerView *)bannerView {
    [_delegate didDismissModal];
}

- (void)adViewWillLeaveApplication:(GADBannerView *)bannerView {
    [_delegate willLeaveApp];
}

#pragma mark - GADAdSizeDelegate methods

- (void)adView:(nonnull GADBannerView *)bannerView
willChangeAdSizeTo:(GADAdSize)size {
    _dfpAdSize = size.size;
}
#pragma mark -

- (void)syncTimetAction{
    if (!self.notified) {
        [self.delegate
         adServerDidWin:_bannerView];
        self.notified = YES;
    }
}

@end

File: OpenBid/DFP/Swift/SampleApp/SampleApp/EventHandlers/DFP/Banner/DFPBannerEventHandler.swift
/*
 * PubMatic Inc. ("PubMatic") CONFIDENTIAL
 * Unpublished Copyright (c) 2006-2019 PubMatic, All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains the property of
 * PubMatic. The intellectual and technical concepts contained herein are
 * proprietary to PubMatic and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material is
 * strictly forbidden unless prior written permission is obtained from PubMatic.
 * Access to the source code contained herein is hereby forbidden to anyone
 * except current PubMatic employees, managers or contractors who have executed
 * Confidentiality and Non-disclosure agreements explicitly covering such
 * access.
 *
 * The copyright notice above does not evidence any actual or intended
 * publication or disclosure  of  this source code, which includes information
 * that is confidential and/or proprietary, and is a trade secret, of  PubMatic.
 * ANY REPRODUCTION, MODIFICATION, DISTRIBUTION, PUBLIC  PERFORMANCE, OR PUBLIC
 * DISPLAY OF OR THROUGH USE  OF THIS  SOURCE CODE  WITHOUT  THE EXPRESS WRITTEN
 * CONSENT OF PubMatic IS STRICTLY PROHIBITED, AND IN VIOLATION OF APPLICABLE
 * LAWS AND INTERNATIONAL TREATIES.  THE RECEIPT OR POSSESSION OF  THIS SOURCE
 * CODE AND/OR RELATED INFORMATION DOES NOT CONVEY OR IMPLY ANY RIGHTS TO
 * REPRODUCE, DISCLOSE OR DISTRIBUTE ITS CONTENTS, OR TO MANUFACTURE, USE, OR
 * SELL ANYTHING THAT IT  MAY DESCRIBE, IN WHOLE OR IN PART.
 */

import UIKit
import GoogleMobileAds

/*!
 This class is responsible for communication between OpenBid banner view and DFP banner view.
 It implements the POBBannerEvent protocol. it notifies event back to OpenBid SDK using POBBannerEventDelegate methods
 */
class DFPBannerEventHandler: NSObject,POBBannerEvent,GADBannerViewDelegate,GADAppEventDelegate,GADAdSizeDelegate {

    let SYNC_TIMEOUT_INTEREVAL = 0.6
        
    var bannerView: DFPBannerView?
    var timer: Timer?
    weak var _delegate: POBBannerEventDelegate?
    var dfpAdSize = CGSize.zero
    var notified = false
    var isAppEventExpected = false
    var adSizes = [NSValue]()
    
    /*!
     A configureBlock that is called before event handler makes ad request call to DFP SDK. It passes DFPBannerView & DFPRequest which will be used to make ad request.
     */
    var configBlock: ((_ view: DFPBannerView, _ request: DFPRequest)->Void)?


    /*!
     Initializes and returns a event handler with the specified DFP ad unit and ad sizes
     
     @param adUnitId DFP ad unit id
     @param validSizes Array of NSValue encoded GADAdSize structs. Never create your own GADAdSize directly. Use one of the predefined
     standard ad sizes (such as kGADAdSizeBanner), or create one using the GADAdSizeFromCGSize
     method.
     Example:
     <pre>
     NSArray *validSizes = @[
     NSValueFromGADAdSize(kGADAdSizeBanner),
     NSValueFromGADAdSize(kGADAdSizeLargeBanner)
     ];
     bannerView.validAdSizes = validSizes;
     </pre>
     
     */
    init(_ adUnitId: String, andSizes validSizes: [NSValue]) {
        super.init()
        bannerView = DFPBannerView()
        bannerView?.adUnitID = adUnitId
        
        adSizes = validSizes
        
        bannerView?.validAdSizes = validSizes
        dfpAdSize = (bannerView?.adSize.size)!
        
        bannerView?.delegate = self
        bannerView?.appEventDelegate = self
        bannerView?.adSizeDelegate = self
    }
    
    deinit {
        bannerView?.delegate = nil
        bannerView = nil
    }
    
    func setDelegate(_ delegate: POBBannerEventDelegate!) {
        _delegate = delegate
    }

    func requestAd(with bid: POBBid?) {
        
        notified = false
        isAppEventExpected = false
        bannerView?.rootViewController = _delegate?.viewControllerForPresentingModal()
        let dfpRequest = DFPRequest()
        
        self.configBlock?(bannerView!,dfpRequest)
        
        if !(bannerView?.appEventDelegate is DFPBannerEventHandler
            && bannerView?.delegate is DFPBannerEventHandler) {
                NSLog("Do not set DFP delegates. These are used by DFPBannerEventHandler internally.");
        }
        
        if bid != nil {
            if bid?.status.boolValue ?? false {
                isAppEventExpected = true
            }
            
            let customTargeting = NSMutableDictionary()
            customTargeting.addEntries(from: dfpRequest.customTargeting ?? [:])
            customTargeting.addEntries(from: bid?.targetingInfo() ?? [:]);
            dfpRequest.customTargeting = customTargeting as? [AnyHashable : Any]
            print("Custom targeting : \(String(describing: customTargeting.description))")
        }
        bannerView?.load(dfpRequest)
    }
    
    func adContentSize() -> CGSize {
        return dfpAdSize
    }
    
    func requestedAdSizes() -> [POBAdSize] {
        var sizes:[POBAdSize] = []
        for value in adSizes {
            let gAdSize = GADAdSizeFromNSValue(value)
            sizes.append(POBAdSize(cgSize: gAdSize.size))
        }
        return sizes
    }
    
    func adView(_ banner: GADBannerView, didReceiveAppEvent name: String, withInfo info: String?) {
        if (name == "pubmaticdm") {
            if notified {
                var userInfo: [AnyHashable : Any]? = nil
                let errorMessage = "App event is called unexpetedly"
                userInfo = [NSLocalizedDescriptionKey: NSLocalizedString(errorMessage, comment: "")]
                let error = NSError(domain: kPOBErrorDomain, code: POBErrorCode.signalingError.rawValue, userInfo: userInfo as? [String : Any])
                _delegate?.failedWithError(error)
                return
            }
            notified = true
            _delegate?.openBidPartnerDidWin()
        }
    }
    
    func adViewDidReceiveAd(_ bannerView: GADBannerView) {
        
        // If already notifed, skip waiting for app event
        if notified {
            return
        }
        
        // If PubMatic have provided non-zero bir price, expect for app event for fixed time interval, otherwise consider as DFP has won & serving its own ad
        if !isAppEventExpected {
            _delegate?.adServerDidWin(bannerView)
            notified = true
        } else {
            // Timer to synchronize did recieve and app event callback as their sequence is not fixed
            timer?.invalidate()
            timer = Timer.scheduledTimer(timeInterval: SYNC_TIMEOUT_INTEREVAL, target: self, selector: #selector(self.syncTimetAction), userInfo: nil, repeats: false)
        }
    }
    
    @objc func syncTimetAction() {
        if !notified {
            _delegate?.adServerDidWin(bannerView)
            notified = true
        }
    }
    
    func adView(_ bannerView: GADBannerView, didFailToReceiveAdWithError error: GADRequestError) {
        
        var eventErr:NSError?
        switch error.code {
        case GADErrorCode.noFill.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorNoAds)
            
        case GADErrorCode.invalidRequest.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorInvalidRequest)
            
        case GADErrorCode.networkError.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorNetworkError)
            
        case GADErrorCode.timeout.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorTimeout)
            
        case GADErrorCode.internalError.rawValue:
            fallthrough
        case GADErrorCode.interstitialAlreadyUsed.rawValue:
            fallthrough
        case GADErrorCode.mediationDataError.rawValue:
            fallthrough
        case GADErrorCode.mediationNoFill.rawValue:
            fallthrough
        case GADErrorCode.mediationAdapterError.rawValue:
            fallthrough
        case GADErrorCode.invalidArgument.rawValue:
            fallthrough
        case GADErrorCode.mediationInvalidAdSize.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorInternalError)
            
        case GADErrorCode.receivedInvalidResponse.rawValue:
            eventErr = eventError(error, code: POBErrorCode.errorInvalidResponse)

        default:
            eventErr = error
            break
        }
        _delegate?.failedWithError(eventErr)
    }
    
    func eventError(_ error : GADRequestError, code : POBErrorCode) -> NSError {
        let eventError = NSError(domain: kPOBErrorDomain, code:code.rawValue, userInfo: error.userInfo)
        return eventError
    }
    
    func adViewWillPresentScreen(_ bannerView: GADBannerView) {
        _delegate?.willPresentModal()
    }

    func adViewDidDismissScreen(_ bannerView: GADBannerView) {
        _delegate?.didDismissModal()
    }
    
    func adViewWillLeaveApplication(_ bannerView: GADBannerView) {
        _delegate?.willLeaveApp()
    }
    
    func adView(_ bannerView: GADBannerView, willChangeAdSizeTo size: GADAdSize) {
        dfpAdSize = size.size
    }


}
